add (and default to on) --preview

tbg opened the pull request:
![demo](https://github.com/user-attachments/assets/7496b809-806a-451d-9af2-98acfecdbc2c)

  With this flag, the benchdiff output is rendered between each
  interleaved run of both the old and new benches (within the current
  package).
  
  Especially for longer benchmark runs, this is helpful to give an idea
  of what the results are likely going to be. In some cases, this may
  prompt the user to call it early, thus saving time. But it's also
  pleasant to be able to just check in, which is my primary motivation
  for this change.

tbg committed:
add BenchmarkDummy

tbg committed:
add (and default to on) --preview

With this flag, the benchdiff output is rendered between each
interleaved run of both the old and new benches (within the current
package).

Especially for longer benchmark runs, this is helpful to give an idea
of what the results are likely going to be. In some cases, this may
prompt the user to call it early, thus saving time. But it's also
pleasant to be able to just check in, which is my primary motivation
for this change.

RaduBerinde commented on main.go:
```diff
@@ -336,17 +339,28 @@ func runCmpBenches(
-			spinner.Update(progress)
+			var buf bytes.Buffer
+			if preview {
+				_, err := processBenchOutput(ctx, &buf, bs1, bs2, true, text, tests, nil)
```
[nit] add /* byName */

RaduBerinde replied:
Should we skip this when `j==0`? Not sure what it would show (maybe just the header columns?)

RaduBerinde reviewed (approved):
Nice!

RaduBerinde commented on main.go:
```diff
@@ -336,17 +339,28 @@ func runCmpBenches(
+				if err != nil {
+					return err
+				}
+				_, _ = fmt.Fprintln(&buf)
```
[supernit] There's an extra space because of how we're using the spinner. Maybe add a `\n` at the beginning of `buf` so it doesn't matter. Or change the spinner to not add the space if `prefix` is empty and use an empty prefix.

tbg committed:
ui: don't force leading space with spinner

This avoids indenting the first line of the incremental benchmark output:

```
running benchmarks:
 name                                                 old time/op    new time/op    delta
Sysbench/parallel=true/SQL/3node/oltp_read_write-24    1.20ms ± 5%    1.20ms ± 6%   ~     (p=0.863 n=9+9)
```

now becomes

```
running benchmarks:
name                                                  old time/op    new time/op    delta
Sysbench/parallel=true/SQL/3node/oltp_read_write-24    1.20ms ± 5%    1.20ms ± 6%   ~     (p=0.863 n=9+9)
```

tbg committed:
skip preview on first iteration

It was printing a newline only anyway, but this is cleaner.

tbg commented:
Thanks Radu! Addressed the comments.

RaduBerinde commented:
This behaves really bad when there are a lot of benchmarks that don't fit the screen. Everything keeps getting output to the screen over and over as the thingy is spinning.

RaduBerinde commented:
It may also have been bad because it was through ssh. I think we should avoid printing out all those lines every 100ms. We should be printing the lines once per iteration and then reprint just the one line with the spinner. I may give it a shot.

tbg commented:
Do you have an example invocation? I use benchdiff over mosh all the time
but haven't seen this issue, though admittedly I always run the same
benchmarks :)

On Sun, Mar 2, 2025, 04:00 RaduBerinde ***@***.***> wrote:

> It may also have been bad because it was through ssh. I think we should
> avoid printing out all those lines every 100ms. We should be printing the
> lines once per iteration and then reprint just the one line with the
> spinner. I may give it a shot.
>
> —
> Reply to this email directly, view it on GitHub
> <https://protect.checkpoint.com/v2/r01/___https://github.com/nvanbenschoten/benchdiff/pull/6%23issuecomment-2692528877___.YzJ1OmNvY2tyb2FjaGxhYnMyOmM6Zzo5ZDVjZjJhNTVlMTk3ZGQ0ZjFiNWI3NzYwNWE3M2U4Zjo3OmI1Mjk6NTkwN2VmNjg3YmZmZjRiYmNiMTg5NjZhZWFlNDhjOGM4ZmQyNTQ4MGY4NTU3MGY0NmRkODM1YTVlY2IwNzU0ODpoOlQ6Tg>,
> or unsubscribe
> <https://protect.checkpoint.com/v2/r01/___https://github.com/notifications/unsubscribe-auth/ABGXPZFDMG2FQJ5M5BDKLP32SJX2TAVCNFSM6AAAAABVEZ56DKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOJSGUZDQOBXG4___.YzJ1OmNvY2tyb2FjaGxhYnMyOmM6Zzo5ZDVjZjJhNTVlMTk3ZGQ0ZjFiNWI3NzYwNWE3M2U4Zjo3OmRmMGE6ZGNhYTRiZDE0YWFmMjNlN2Q1N2RkMmI1ZDIwODNhMWI3ODRhYzhhM2I2OGFkNzE4NGE0NGZmMGU3MzAyMmJiNTpoOlQ6Tg>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
> [image: RaduBerinde]*RaduBerinde* left a comment
> (nvanbenschoten/benchdiff#6)
> <https://protect.checkpoint.com/v2/r01/___https://github.com/nvanbenschoten/benchdiff/pull/6%23issuecomment-2692528877___.YzJ1OmNvY2tyb2FjaGxhYnMyOmM6Zzo5ZDVjZjJhNTVlMTk3ZGQ0ZjFiNWI3NzYwNWE3M2U4Zjo3OjYyMDk6OTQ5YWU4MzZmZDlmYWYzMDc4Zjc2NjQ1YjU0MThhMzg2ZTdhMjYyMzI4NzZkYzViMTlhOTAxN2JiMzE3YzljZjpoOlQ6Tg>
>
> It may also have been bad because it was through ssh. I think we should
> avoid printing out all those lines every 100ms. We should be printing the
> lines once per iteration and then reprint just the one line with the
> spinner. I may give it a shot.
>
> —
> Reply to this email directly, view it on GitHub
> <https://protect.checkpoint.com/v2/r01/___https://github.com/nvanbenschoten/benchdiff/pull/6%23issuecomment-2692528877___.YzJ1OmNvY2tyb2FjaGxhYnMyOmM6Zzo5ZDVjZjJhNTVlMTk3ZGQ0ZjFiNWI3NzYwNWE3M2U4Zjo3OjFhNTM6YWVhMDM0ZDNhZmFlNzY4MjE0MTE0OThlMGI4MmM3NGJlNTNmOWI2OGVhNzIxMjAxODk4NzZmNmRkZGJmOWJkMTpoOlQ6Tg>,
> or unsubscribe
> <https://protect.checkpoint.com/v2/r01/___https://github.com/notifications/unsubscribe-auth/ABGXPZFDMG2FQJ5M5BDKLP32SJX2TAVCNFSM6AAAAABVEZ56DKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOJSGUZDQOBXG4___.YzJ1OmNvY2tyb2FjaGxhYnMyOmM6Zzo5ZDVjZjJhNTVlMTk3ZGQ0ZjFiNWI3NzYwNWE3M2U4Zjo3OjZkYTc6NWI1ZGNiZDBkYzM2NmMwOTY4MDFhMzZkMWU5YjA1YTIzYzZjZDYzMjRhYzE3OTk4MThhMjIxYWRiMTI1NWEzMjpoOlQ6Tg>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>

RaduBerinde commented:
https://github.com/nvanbenschoten/benchdiff/pull/13
